haystack/pipelines/base.py

```diff
@@ -373,12 +390,17 @@ def add_node(self, component: BaseComponent, name: str, inputs: List[str]):
                 )
             self.graph = _init_pipeline_graph(root_node_name=candidate_roots[0])
 
-        component_definitions = get_component_definitions(pipeline_config=self.get_config())
-
-        # Check for duplicates before adding the definition
-        if name in component_definitions.keys():
+        # Check for duplicate names before adding the component
+        # Note that the very same component must be addable multiple times:
+        # E.g. for indexing pipelines it's common to add a retriever first and a document store afterwards.
+        # The document store is already being used by the retriever however.
+        # Thus the very same document store will be added twice, first as a subcomponent of the retriever and second as a first level node.
+        if name in self.components and self.components[name] != component:
```

Comment:
Implicit iteration on dict keys make `self.components` look like a list. Let's change it to `if name in self.components.keys()`